src/runtime/hexagon/ops/conv2d_hvx.cc

```diff
@@ -0,0 +1,473 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <HAP_compute_res.h>
+#include <hexagon_types.h>
+#include <hvx_hexagon_protos.h>
+#include <tvm/runtime/c_runtime_api.h>
+#include <tvm/runtime/device_api.h>
+
+#include <algorithm>
+#include <cassert>
+#include <cinttypes>
+
+#include "tvm/runtime/hexagon/ops/conv2d.h"
+
+// Current limitations:
+// - N in NHWC must be 1
+// - dilated convolutions are not supported
+// - Bias is not accepted
+// - Optional "relu" is not performed
+
+// Packed arguments:
+//   0: DLTensor activations (NHWC)
+//   1: DLTensor weights (HWIO)
+//   2: int offset_top
+//   3: int offset_left
+//   4: int stride_h
+//   5: int stride_w
+//   6: DLTensor output (NHWC)
+extern "C" int conv2d_packed(TVMValue* args, int* type_codes, int num_args, TVMValue* out_val,
+                             int out_code, void* res_handle);
+
+namespace tvm {
+namespace runtime {
+namespace hexagon {
+
+inline uint16_t* getElementPtr(int block_out_y, int block_out_x, int block_out_c, int yi, int xio,
```

Comment:
If `uint16_t` is a stand-in for qfloat16, is there a typename supplied by the Hexagon SDK that might be more descriptive?  

This might not be a valid example, but I'm thinking of something _like_ `qhl_q0_t`.

Of if the SDK provides no such type name, could it make sense to just define one here?  E.g., `using qfloat16_ptr = uint16_t*;`

Reply:
It's not qfloat16, the storage type is _Float16.  I'm opposed to using such typedefs, since they actually obscure things instead of clarifying anything.

Reply:
Any reason we couldn't just replace `uint16_t` with `_Float16` then?  AFAICT both GCC and LLVM recognize `_Float16`, and IIUC we're currently only planning to ever build this code with LLVM.

Reply:
The recognizing of `_Float16` in clang at least is target-specific.  For Hexagon, clang recognizes it when the target architecture is >= v68.